### PR TITLE
Bare `Tuple` should mean `Tuple[Any, ...]`.

### LINKED ISSUE #2184: Change meaning of unparameterized Tuple to Tuple[Any, ...] (== tuple)
Currently, `Type` (without parameters) means a 0-length tuple, IOW `Tuple[()]` (which is the PEP 484 way to spell it). Analogy with `List` suggests it should be an unconstrained tuple, i.e. `Tuple` == `tuple` == `Tuple[Any, ...]`. The second half is already true. The `List` analogy is `List` == `list` == `List[Any]`. (The last part of that is spelled differently, but is a closer analogy to `Tuple[Any, ...]` than to `Tuple[Any]`, since the latter is constrained to 1-length tuples.)

### PR DESCRIPTION
**Pull request by gvanrossum**
Fix #2184. (See also python/typing#284.)

### CODE REVIEW

File: test-data/unit/check-tuples.test [hunk 1]
```diff
@@ -820,3 +820,20 @@ from typing import List, Tuple
 def f() -> Tuple[List[str], ...]:
     return ([],)
 [builtins fixtures/tuple.pyi]
+
+[case testTupleWithoutArgs]
+from typing import Tuple
+def f(a: Tuple) -> None: pass
+f(())
+f((1,))
+f(('', ''))
```
Comment (JukkaL) [inline test-data/unit/check-tuples.test]: Call with a non-tuple argument such as an int.

**Comment by gvanrossum** [comment]
Ouch. I just realized that I broke the behavior of `Tuple[()]`. I added a test for that but didn't finish that thought.

**Comment by gvanrossum** [comment]
@JukkaL Another look? I added the test with int you requested, and fixed the issue I introduced with `Tuple[()]` -- that was a little awkward since the information about whether the original code was `Tuple` or `Tuple[()]` was lost by the time it reached typeanal.py.

I also looked into the difference with the next block and decided to leave that alone (although reducing it to `self.builtin_type('builtins.tuple', [t.args[0].accept(self)])` made all tests pass, I decided to give it the benefit of the doubt, maybe the line number is important after all).

**Comment by elazarg** [comment]
If I understand correctly, this flag may be useful in order to write `Union[]` as a synonym for `UninhabitedType` (i.e. NoReturn) , while still allowing `Union` to be `Union[Any]` (which is actually `Any`).

**Comment by gvanrossum** [comment]
Sorry, Union[] (like Tuple[]) is invalid syntax. We use Tuple[()] as an
escape, but Union[()] would be too ugly for such a purpose.

File: test-data/unit/check-tuples.test [hunk 2]
```diff
@@ -820,3 +820,22 @@ from typing import List, Tuple
...
+f(0)  # E: Argument 1 to "f" has incompatible type "int"; expected Tuple[Any, ...]
+[builtins fixtures/tuple.pyi]
+
+[case testTupleSingleton]
+# flags: --fast-parser
+from typing import Tuple
+def f(a: Tuple[()]) -> None: pass
+f(())
+f((1,))  # E: Argument 1 to "f" has incompatible type "Tuple[int]"; expected "Tuple[]"
```
Comment (JukkaL) [inline test-data/unit/check-tuples.test]: Maybe update the output of an empty tuple to be `Tuple[()]` instead of `Tuple[]`?

**Comment by JukkaL** [comment]
Other than the comment above, this looks good to me.

**Comment by gvanrossum** [comment]
About Tuple[] vs. Tuple[()] in output... I have the patch for that ready,
but I'm not sure I like it. I had to change some other tests that had
Tuple[] in their output, and somehow I have a feeling that for a user,
*seeing* Tuple[] is more indicative of "the type of an empty tuple" than
Tuple[()], which only works for obscure reasons. So I think I'd like to
leave this for a later time.

**Comment by JukkaL** [comment]
OK, sounds reasonable `Tuple[()]` really isn't the most intuitive thing ever.

**Pull request merged**